content/blog/2019-12-02-equivalence-modulo-inputs/index.md

```diff
@@ -0,0 +1,400 @@
+that GCC Tree Optimization and RTL optimization were the components with the
+most discovered bugs (LLVM developers did not classify the reported bugs). These
+components did not necessarily have more bugs than others, nor were these the
+only possible bugs.
+
+The authors do not make an attempt to evaluate the search space that Orion
+explored in producing these reported bugs. Nor do they explicitly determine the
+proportion of the generated variants that led to identified bugs. They only
+report that they didn't record how many seed programs they started with or how
+many variants they generated (merely estimating "millions to tens of millions").
+They also do not report (and likely did not record) the Csmith configurations or
+Orion's dynamic pruning parameters.
+
+#### Qualitative examples
+
+The confirmed bugs are said to span compiler segfaults, internal compiler
+errors, performance problems, and wrong code generation. The authors present and
+interpret a handful of the bugs that were confirmed and fixed by compiler
+developers. We highlight just two of those for a flavor of the generated
+programs. Note that the authors only show the reduced code they reported to
+compiler developers; they show neither the non-reduced versions nor the EMI
+variants.
+
+The following example led to a segfault when compiled with GCC due to a wrong
+offset computation in an optimization pass called "predictive commoning," a form
+of common subexpression elimination:
+
+```c
+int b, f, d[5][2];
+unsigned int c;
+int main() {
+  for (c = 0; c < 2; c++)
+    if (d[b + 3][c] & d[b + 4][c])
+      if (f)
+        break;
+  return 0;
+}
+```
+
+Clang incorrectly vectorized the following code:
+
+```c
+int main() {
+  int a = 1;
+  char b = 0;
+  lbl:
+    a &= 4;
+    b--;
+    if (b) goto lbl;
+  return a;
+}
+```
+
+#### Current statistics
+
+The website for the [EMI project][emi] shows the number of bugs found and fixed
+by tools using the EMI methodology.
+It shows an astronomical number of bugs found in GCC and LLVM,
+and the usefulness of the technique for compilers for other languages
+like Scala.
+
+Compiler    | Bugs reported | Bugs fixed
+------------|---------------|-----------
+GCC/LLVM    | 1,602         | 1,007
+ICC         | 35            | ?
+CompCert    | 31            | 27
+scala/dotty | 42            | 17
+
+
+[emi]: https://web.cs.ucdavis.edu/~su/emi-project/
+
+
+
+## Discussion
+
+The remaining examples in the paper cover problems with jump-threading logic,
+global value numbering, inlining, vectorization, and performance. Because the
+authors analyze only a few cherry-picked examples, a question remains: Are these
+eight examples representative of all of the other bugs?
+
+Additionally, the authors claim: "EMI variants generated from existing code, say
+via Orion, are likely programs that people actually write."
+Is this true, especially when random programs are used as seeds?
```

Comment:
Agreed; this seems like a pretty dubious claim by itself. The results in the evaluation seem to suggest that EMI is most useful as an "amplifier" for an existing fuzzer like CSmith—not as some sort of standalone tool.